File: backend/app/ledger/merkle.py
"""Merkle tree with SHA-256 leaves and inclusion proofs.

This is the cryptographic core of LedgerRAG. It lets us commit to an entire set
of document chunks with a single root hash, and later prove that any one chunk
belongs to that set using an O(log n) inclusion proof -- without revealing the
other chunks.

Design notes:
- Leaves are hashed with a 0x00 prefix, internal nodes with 0x01, to prevent
  second-preimage attacks (a known Merkle-tree pitfall).
- Odd nodes at a level are promoted (duplicated) to the next level.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Literal

LEAF_PREFIX = b"\x00"
NODE_PREFIX = b"\x01"


def _sha256(data: bytes) -> bytes:
    return hashlib.sha256(data).digest()


def hash_leaf(data: bytes) -> bytes:
    """Hash a leaf value (a chunk's normalized bytes)."""
    return _sha256(LEAF_PREFIX + data)


def hash_nodes(left: bytes, right: bytes) -> bytes:
    """Hash two child nodes into their parent."""
    return _sha256(NODE_PREFIX + left + right)


@dataclass(frozen=True)
class ProofStep:
    """One step in a Merkle inclusion proof: a sibling hash and its side."""

    sibling: bytes
    side: Literal["left", "right"]  # which side the sibling is on


class MerkleTree:
    """An immutable Merkle tree built from a list of leaf byte-strings."""

    def __init__(self, leaves: list[bytes]):
        if not leaves:
            raise ValueError("MerkleTree requires at least one leaf")
        self._leaf_hashes: list[bytes] = [hash_leaf(x) for x in leaves]
        self._levels: list[list[bytes]] = self._build(self._leaf_hashes)

    @staticmethod
    def _build(leaf_hashes: list[bytes]) -> list[list[bytes]]:
        levels = [leaf_hashes]
        current = leaf_hashes
        while len(current) > 1:
            nxt: list[bytes] = []
            for i in range(0, len(current), 2):
                left = current[i]
                right = current[i + 1] if i + 1 < len(current) else current[i]
                nxt.append(hash_nodes(left, right))
            levels.append(nxt)
            current = nxt
        return levels

    @property
    def root(self) -> bytes:
        return self._levels[-1][0]

    @property
    def leaf_count(self) -> int:
        return len(self._leaf_hashes)

    def leaf_hash(self, index: int) -> bytes:
        return self._leaf_hashes[index]

    def inclusion_proof(self, index: int) -> list[ProofStep]:
        """Return the Merkle path proving leaf `index` is under `root`."""
        if not 0 <= index < self.leaf_count:
            raise IndexError("leaf index out of range")
        proof: list[ProofStep] = []
        idx = index
        for level in self._levels[:-1]:  # every level except the root
            is_right = idx % 2 == 1
            if is_right:
                sibling_idx = idx - 1
                side: Literal["left", "right"] = "left"
            else:
                sibling_idx = idx + 1 if idx + 1 < len(level) else idx
                side = "right"
            proof.append(ProofStep(sibling=level[sibling_idx], side=side))
            idx //= 2
        return proof


def verify_inclusion(leaf_data: bytes, proof: list[ProofStep], root: bytes) -> bool:
    """Independently verify that `leaf_data` is committed under `root`.

    This function is the whole point of LedgerRAG's verifiability: given only the
    leaf bytes, the proof, and a (signed) root, anyone can confirm membership
    without trusting the server that produced the answer.
    """
    computed = hash_leaf(leaf_data)
    for step in proof:
        if step.side == "left":
            computed = hash_nodes(step.sibling, computed)
        else:
            computed = hash_nodes(computed, step.sibling)
    return computed == root

File: go/merkle.go
// Package ledgerrag is a cross-language port of LedgerRAG's cryptographic core:
// a SHA-256 Merkle tree with inclusion proofs and Ed25519-signed roots, plus the
// independent verifier that re-checks a proof-carrying answer without trusting
// the server that produced it.
//
// The hashing and signing rules are reproduced exactly from the Python
// reference so a proof generated in Python (or C#/Java) verifies here too:
//
//   - leaf hash = SHA-256(0x00 || utf8(text))
//   - node hash = SHA-256(0x01 || left || right)
//   - root msg  = root || prev_root(optional) || leaf_count (8 bytes, big-endian)
//   - signature = Ed25519 over root msg, verified with the raw public key
package ledgerrag

import (
	"bytes"
	"crypto/sha256"
	"errors"
)

// leafPrefix and nodePrefix domain-separate leaves from internal nodes so a leaf
// hash can never be reinterpreted as an internal node (second-preimage defence).
var (
	leafPrefix = []byte{0x00}
	nodePrefix = []byte{0x01}
)

// HashLeaf hashes a leaf value (a chunk's normalized bytes) with the 0x00 prefix.
func HashLeaf(data []byte) []byte {
	h := sha256.New()
	h.Write(leafPrefix)
	h.Write(data)
	return h.Sum(nil)
}

// HashNodes hashes two child nodes into their parent with the 0x01 prefix.
func HashNodes(left, right []byte) []byte {
	h := sha256.New()
	h.Write(nodePrefix)
	h.Write(left)
	h.Write(right)
	return h.Sum(nil)
}

// ProofStep is one step in a Merkle inclusion proof: a sibling hash and the side
// it sits on ("left" or "right").
type ProofStep struct {
	Sibling []byte
	Side    string
}

// MerkleTree is an immutable Merkle tree built from a list of leaf byte-strings.
type MerkleTree struct {
	leafHashes [][]byte
	levels     [][][]byte
}

// NewMerkleTree builds a Merkle tree over leaves. It returns an error when the
// leaf list is empty.
func NewMerkleTree(leaves [][]byte) (*MerkleTree, error) {
	if len(leaves) == 0 {
		return nil, errors.New("MerkleTree requires at least one leaf")
	}
	leafHashes := make([][]byte, len(leaves))
	for i, x := range leaves {
		leafHashes[i] = HashLeaf(x)
	}
	return &MerkleTree{leafHashes: leafHashes, levels: buildLevels(leafHashes)}, nil
}

// buildLevels folds the leaf hashes up to the root, promoting (duplicating) the
// last node when a level has an odd number of entries.
func buildLevels(leafHashes [][]byte) [][][]byte {
	levels := [][][]byte{leafHashes}
	current := leafHashes
	for len(current) > 1 {
		next := make([][]byte, 0, (len(current)+1)/2)
		for i := 0; i < len(current); i += 2 {
			left := current[i]
			var right []byte
			if i+1 < len(current) {
				right = current[i+1]
			} else {
				right = current[i]
			}
			next = append(next, HashNodes(left, right))
		}
		levels = append(levels, next)
		current = next
	}
	return levels
}

// Root returns the Merkle root hash.
func (t *MerkleTree) Root() []byte {
	return t.levels[len(t.levels)-1][0]
}

// LeafCount returns the number of leaves committed by the tree.
func (t *MerkleTree) LeafCount() int {
	return len(t.leafHashes)
}

// LeafHash returns the (0x00-prefixed) hash of the leaf at index.
func (t *MerkleTree) LeafHash(index int) []byte {
	return t.leafHashes[index]
}

// InclusionProof returns the Merkle path proving the leaf at index is under Root.
func (t *MerkleTree) InclusionProof(index int) ([]ProofStep, error) {
	if index < 0 || index >= t.LeafCount() {
		return nil, errors.New("leaf index out of range")
	}
	proof := []ProofStep{}
	idx := index
	for l := 0; l < len(t.levels)-1; l++ { // every level except the root
		level := t.levels[l]
		var siblingIdx int
		var side string
		if idx%2 == 1 {
			siblingIdx = idx - 1
			side = "left"
		} else {
			if idx+1 < len(level) {
				siblingIdx = idx + 1
			} else {
				siblingIdx = idx
			}
			side = "right"
		}
		proof = append(proof, ProofStep{Sibling: level[siblingIdx], Side: side})
		idx /= 2
	}
	return proof, nil
}

// VerifyInclusion independently confirms leafData is committed under root, given
// only the leaf bytes, the proof, and a (trusted) root.
func VerifyInclusion(leafData []byte, proof []ProofStep, root []byte) bool {
	computed := HashLeaf(leafData)
	for _, step := range proof {
		if step.Side == "left" {
			computed = HashNodes(step.Sibling, computed)
		} else {
			computed = HashNodes(computed, step.Sibling)
		}
	}
	return bytes.Equal(computed, root)
}
